[Merged by Bors] - feat(MeasureTheory/Function/ConditionalExpectation): relax `integral_condexp` assumption and add `integral_condexp_indicator`

gaetanserre opened the pull request:
Remove the `Integrable` assumption of `integral_condexp` and add the total probability law using `condexp` as conditional probability.


---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

RemyDegenne commented on Mathlib/MeasureTheory/Function/ConditionalExpectation/Basic.lean:
```diff
@@ -420,4 +420,48 @@ theorem tendsto_condexp_unique (fs gs : ℕ → α → F') (f g : α → F')
+        lintegral_indicator (fun _ ↦ 1) hAm0
+      _ = μ A := set_lintegral_one A
+      _ < ⊤ := measure_lt_top μ A
+    have hAEmα : AEStronglyMeasurable (A.indicator (fun (_ : α) ↦ (1 : ℝ))) μ := by
+      suffices StronglyMeasurable (A.indicator (fun (_ : α) ↦ (1 : ℝ))) by
+        exact StronglyMeasurable.aestronglyMeasurable this
+      exact StronglyMeasurable.mono
+        (StronglyMeasurable.indicator stronglyMeasurable_const hA)
+        (measurable_iff_comap_le.mp hY)
```
Exactly the same proof, written in a compact way:
```suggestion
    have hAEmα : AEStronglyMeasurable (A.indicator (fun _ ↦ (1 : ℝ))) μ :=
      ((stronglyMeasurable_const.indicator hA).mono hY.comap_le).aestronglyMeasurable
```

RemyDegenne commented on Mathlib/MeasureTheory/Function/ConditionalExpectation/Basic.lean:
```diff
@@ -420,4 +420,48 @@ theorem tendsto_condexp_unique (fs gs : ℕ → α → F') (f g : α → F')
   exact tendsto_nhds_unique_of_eventuallyEq hcond_gs hcond_fs (eventually_of_forall hn_eq)
 #align measure_theory.tendsto_condexp_unique MeasureTheory.tendsto_condexp_unique
 
+open Set
+
```
It saves you a single `Set.` in the code. That does not seem worth it. You can event replace that `Set.indicator A` by `A.indicator`.
```suggestion
```

RemyDegenne commented on Mathlib/MeasureTheory/Function/ConditionalExpectation/Basic.lean:
```diff
@@ -420,4 +420,48 @@ theorem tendsto_condexp_unique (fs gs : ℕ → α → F') (f g : α → F')
+
+/-- Total probability law using `condexp` as conditional probability. -/
+theorem total_law_condexp [mF : MeasurableSpace F] {Y : α → F} (hY : Measurable Y)
+    [IsProbabilityMeasure μ] {A : Set α} (hA : MeasurableSet[mF.comap Y] A)
```
Can you extend the result to a finite measure instead of probability measure?

RemyDegenne reviewed (commented):
Thanks!

RemyDegenne commented:
Actually instead of applying all the golfing suggestions above, here is a short proof of your lemma:
```
/-- Total probability law using `condexp` as conditional probability. -/
theorem total_law_condexp [mF : MeasurableSpace F] {Y : α → F} (hY : Measurable Y)
    [IsProbabilityMeasure μ] {A : Set α} (hA : MeasurableSet[mF.comap Y] A) :
    (μ A).toReal = ∫ x, (μ[(A.indicator fun _ ↦ (1 : ℝ)) | mF.comap Y]) x ∂μ := by
  have hAm0 : MeasurableSet A := hY.comap_le A hA
  have total_exp : ∫ x, (μ[A.indicator fun _ ↦ (1 : ℝ)|mF.comap Y]) x ∂μ
      = ∫ x, A.indicator (fun x ↦ 1) x ∂ μ :=
    integral_condexp (measurable_iff_comap_le.mp hY) ((integrable_const _).indicator hAm0)
  rw [total_exp, integral_indicator hAm0, setIntegral_const, smul_eq_mul, mul_one]
```
(Please still read my comments about the name and the assumptions)

JasonKYi commented:
This lemma seems strange to me as you assume `MeasurableSet[mF.comap Y] A` and so the conditional expectation does not do anything to the indicator. Perhaps you only want ``MeasurableSet A`?

gaetanserre commented:
Thank you @RemyDegenne  for all your suggestions. I tried used your shorter proof and changed the name. I have taken note of all previous suggestions for future proofs, thanks.
@JasonKYi: Indeed, I changed the hypothesis to only `MeasurableSet A`.

RemyDegenne commented:
Please check every comment above and mark as resolved all those that you addressed. If something remains (like my comment about switching the two sides of the equality in the result of the lemma, or my question about finite measures), please address those.

gaetanserre commented:
Sorry for that. I think it's done now.

JasonKYi commented:
`integral_condexp` can be generalized so that you don't need integrability. Namely 
```lean
lemma foo {α : Type u_1} {F' : Type u_3}
  [NormedAddCommGroup F'] [NormedSpace ℝ F'] [CompleteSpace F']
  {m : MeasurableSpace α} {m0 : MeasurableSpace α} {μ : MeasureTheory.Measure α}
  (f : α → F') (hm : m ≤ m0) [hμm : MeasureTheory.SigmaFinite (μ.trim hm)] :
∫ (x : α), MeasureTheory.condexp m μ f x ∂μ = ∫ (x : α), f x ∂μ := by
  by_cases hf : MeasureTheory.Integrable f μ
  · exact integral_condexp hm hf
  · simp only [condexp_undef hf, Pi.zero_apply, integral_zero, integral_undef hf]
```
Can you generalize `integral_condexp` and then you can require sigma finite instead of finite measure.

RemyDegenne commented on Mathlib/MeasureTheory/Function/ConditionalExpectation/Basic.lean:
```diff
@@ -420,4 +420,10 @@ theorem tendsto_condexp_unique (fs gs : ℕ → α → F') (f g : α → F')
 
+/-- Total probability law using `condexp` as conditional probability. -/
+theorem integral_condexp_indicator [mF : MeasurableSpace F] {Y : α → F} (hY : Measurable Y)
+    [SigmaFinite (μ.trim (measurable_iff_comap_le.mp hY))] {A : Set α} (hA : MeasurableSet A) :
```
```suggestion
    [SigmaFinite (μ.trim hY.comap_le)] {A : Set α} (hA : MeasurableSet A) :
```

RemyDegenne commented on Mathlib/MeasureTheory/Function/ConditionalExpectation/Basic.lean:
```diff
@@ -420,4 +420,10 @@ theorem tendsto_condexp_unique (fs gs : ℕ → α → F') (f g : α → F')
   exact tendsto_nhds_unique_of_eventuallyEq hcond_gs hcond_fs (eventually_of_forall hn_eq)
 #align measure_theory.tendsto_condexp_unique MeasureTheory.tendsto_condexp_unique
 
+/-- Total probability law using `condexp` as conditional probability. -/
+theorem integral_condexp_indicator [mF : MeasurableSpace F] {Y : α → F} (hY : Measurable Y)
+    [SigmaFinite (μ.trim hY.comap_le)] {A : Set α} (hA : MeasurableSet A) :
+    ∫ x, (μ[(A.indicator fun _ ↦ (1 : ℝ)) | mF.comap Y]) x ∂μ = (μ A).toReal := by
+  rw [integral_condexp, integral_indicator hA, setIntegral_const, smul_eq_mul, mul_one]
```
Can you put this right after `integral_condexp`? Since it's a consequence of that lemma, they should go together.

RemyDegenne commented:
> Thank you! I did it and modified few proofs consequently. However, I'm wondering if it's possible to have a "nicer" assumption than `[SigmaFinite (μ.trim (measurable_iff_comap_le.mp hY))]`.

Sigma-finite with respect to the sub-sigma-algebra is the minimal assumption under which conditional expectations exist, if I recall correctly. The book Analysis in Banach Spaces, by Hytönen et al, is a good reference.

If your measure is finite, then the typeclass inference system will infer that for you, so the fact that it is somewhat complicated does not matter much.

gaetanserre commented:
> Sigma-finite with respect to the sub-sigma-algebra is the minimal assumption under which conditional expectations exist, if I recall correctly. The book Analysis in Banach Spaces, by Hytönen et al, is a good reference.
Ok I see. Indeed, it seems to be a fair assumption. I'll check this reference, thanks!

> If your measure is finite, then the typeclass inference system will infer that for you, so the fact that it is somewhat complicated does not matter much.
Indeed, it should be goof for most classical usage I guess.

JasonKYi commented on Mathlib/MeasureTheory/Function/ConditionalExpectation/Basic.lean:
```diff
@@ -226,13 +226,21 @@ theorem setIntegral_condexp (hm : m ≤ m0) [SigmaFinite (μ.trim hm)] (hf : Int
-  suffices ∫ x in Set.univ, (μ[f|m]) x ∂μ = ∫ x in Set.univ, f x ∂μ by
-    simp_rw [integral_univ] at this; exact this
-  exact setIntegral_condexp hm hf (@MeasurableSet.univ _ m)
+  by_cases hf : MeasureTheory.Integrable f μ
```
```suggestion
  by_cases hf : Integrable f μ
```

JasonKYi commented:
Thanks! 
maintainer merge

sgouezel commented:
bors r+
